crates/error_macros/src/lib.rs

```diff
@@ -1,33 +1,145 @@
+extern "C" {
+    fn write(fd: i32, buf: *const u8, count: usize) -> isize;
+    fn exit(status: i32) -> !;
+}
+
+#[cfg(unix)]
+const STDERR_FD: i32 = 2;
+
+#[cfg(windows)]
+extern "C" {
+    fn GetStdHandle(nStdHandle: i32) -> *mut core::ffi::c_void;
+    fn WriteFile(
+        hFile: *mut core::ffi::c_void,
+        lpBuffer: *const u8,
+        nNumberOfBytesToWrite: u32,
+        lpNumberOfBytesWritten: *mut u32,
+        lpOverlapped: *mut core::ffi::c_void,
+    ) -> i32;
+    fn ExitProcess(exit_code: u32) -> !;
+}
+
+#[cfg(windows)]
+const STD_ERROR_HANDLE: i32 = -12;
+
+/// Print each of the given strings to stderr (if it's available; on wasm, nothing will
+/// be printed) and then immediately exit the program with an error.
+/// On wasm, this will trap, and on UNIX or Windows it will exit with a code of 1.
+#[inline(never)]
+#[cold]
+#[cfg(any(unix, windows, target_arch = "wasm32"))]
+pub fn error_and_exit(args: fmt::Arguments) -> ! {
+    use fmt::Write;
+
+    struct StderrWriter;
+
+    impl Write for StderrWriter {
+        #[cfg(unix)]
+        fn write_str(&mut self, s: &str) -> fmt::Result {
+            unsafe {
+                let _ = write(STDERR_FD, s.as_ptr(), s.len());
+            }
+            Ok(())
+        }
+
+        #[cfg(windows)]
+        fn write_str(&mut self, s: &str) -> fmt::Result {
+            unsafe {
+                let handle = GetStdHandle(STD_ERROR_HANDLE);
+                let mut written = 0;
+                let _ = WriteFile(
+                    handle,
+                    s.as_ptr(),
+                    s.len() as u32,
+                    &mut written,
+                    core::ptr::null_mut(),
+                );
+            }
+            Ok(())
+        }
+
+        #[cfg(target_arch = "wasm32")]
+        fn write_str(&mut self, _s: &str) -> fmt::Result {
+            Ok(())
+        }
+    }
+
+    let _ = fmt::write(&mut StderrWriter, args);
+
+    // Write a newline at the end to make sure stderr gets flushed.
+    let _ = StderrWriter.write_str("\n");
+
+    #[cfg(unix)]
+    unsafe {
+        exit(1)
+    }
+
+    #[cfg(windows)]
+    unsafe {
+        ExitProcess(1)
+    }
+
+    #[cfg(target_arch = "wasm32")]
+    {
+        // We have no way to write to any stderr equivalent in wasm,
+        // so just trap to end the program immediately.
+        core::arch::wasm32::unreachable()
+    }
+}
+
+pub const INTERNAL_ERROR_MESSAGE: &str = concat!(
+    "An internal compiler expectation was broken.\n",
+    "This is definitely a compiler bug.\n",
+    "Please file an issue here: <https://github.com/roc-lang/roc/issues/new/choose>\n",
+);
+
 /// `internal_error!` should be used whenever a compiler invariant is broken.
-/// It is a wrapper around panic that tells the user to file a bug.
+/// It tells the user to file a bug and then exits the program with a nonzero exit code.
+/// (On wasm it doesn't tell the user anything, since we don't necessarily have a way to print.)
```

Comment:
You mention here we don't necessarily have a way to print in WASM, but assert above that it's impossible in the comment and the code. Are you leaving the door open for WASI or something? Unless you want to try optimistically printing to Stderr via WASI when possible, you may want to avoid confusion by just saying "we don't have a way to print".